waspc/src/Wasp/Generator/WriteFileDrafts.hs

```diff
@@ -0,0 +1,161 @@
+{-# LANGUAGE TupleSections #-}
+
+module Wasp.Generator.WriteFileDrafts
+  ( writeFileDrafts,
+    fileDraftsToWriteAndFilesToDelete, -- Exported for testing.
+    removeFromChecksumFile,
+  )
+where
+
+import Control.Arrow (Arrow (first))
+import Control.Monad (unless)
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Encode.Pretty as AesonPretty
+import qualified Data.ByteString.Lazy as BSL
+import Data.Either (lefts, rights)
+import Data.List (find)
+import Data.Maybe (isJust, isNothing)
+import StrongPath (Abs, Dir, File', Path', Rel, relfile, (</>))
+import qualified StrongPath as SP
+import System.Directory (removeDirectoryRecursive, removeFile)
+import System.IO.Error (isDoesNotExistError)
+import UnliftIO.Exception (catch, throwIO)
+import Wasp.Generator.Common (ProjectRootDir)
+import Wasp.Generator.FileDraft (FileDraft, write)
+import Wasp.Generator.FileDraft.Writeable (FileOrDirPathRelativeTo, Writeable (getChecksum, getDstPath))
+import Wasp.Util (Checksum)
+
+-- | Writes given file drafts to disk, in the provided destination directory.
+-- Also makes sure to remove any redundant file drafts that have been left on the disk from before.
+-- It is smart when writing, so it doesn't write file drafts that are already written on the disk from before.
+writeFileDrafts :: Path' Abs (Dir ProjectRootDir) -> [FileDraft] -> IO ()
+writeFileDrafts dstDir fileDrafts = do
+  maybePathsToChecksums <- readChecksumFile dstDir
+  case maybePathsToChecksums of
+    -- If checksums file is missing/corrupted, we delete all of the generated code
+    -- since we are not sure what is the state on the disk and don't want redundant files
+    -- to remain after we write the new file drafts.
+    -- Normally we delete redundant left-over files based on the checksum file, but since we don't have it,
+    -- the only way to be sure is just to delete everything.
+    Nothing -> do
+      removeDirectoryRecursive (SP.fromAbsDir dstDir)
+        `catch` \e -> unless (isDoesNotExistError e) $ throwIO e
+    -- We remove checksum file even if it is valid, so in case we fail in the middle writing on the disk
```

Comment:
nice! :D Now whoever reads our source code will successfully pass the interview :D